### PR TITLE
feat: Edit Collection & NFT UI

### LINKED ISSUE #5286: Collection edit UI
we will have a new ui for editing/customizing collection for owners - mainly because introducing big banner on collection detail 

# Figma 
- collection detail handoff 
- https://www.figma.com/file/7BWdGiowrvTucpSFLE0kRX/Collection-detail-handoff?node-id=0%3A1&t=cKi2EVd3pPHjX4n4-1

# Main
image
- it's quite similar to the sidebar for wallet connect - its only sharing the style but position is in the middle of a screen as popup
- Collection owner can get there once he click through more options and then select "edit collection" 
image

## Icon 
- Owner can choose to change collection logo/icon - the current one would be displayed in solid stroke 
image
- if there is not anything filled, it looks like this:
image

## Banner
image
- you can drop your files into the field or select file. 
- here are the states for hovering over or dragging file over it
image

- question for you, should we include supported types? If yes, what are they for banner? the main ones png, jpg etc

## Social links
image
- option for the owner to link certain links for the collection 
- links that are filled out will be shown on the page among social links buttons - left side
- active state for the input is in the figma as well - its the same as with price input on explore

## Save changes 
- primary button - hover effect to bg color

I think we could maybe expand the options in future - something like name edit etc, but for now I think it works Thank you!

### LINKED ISSUE #11099: Support Edits After Minting
Filippo is requesting to add feature to edit the NFT: collection id and items to edit title, description, and property/traits

### He suggested a solution:
Create a new IPFS instance and then link it to the item/collection (if you want to change collection picture)

Open for discussion: Is it possible? Yes, no? Why? How easy it it?

### LINKED ISSUE #11145: Collection Edit: Support Image Update
Add functionality to the collection edit modal allowing users to update the collection image. Ensure that the updated image is reflected on-chain and displayed correctly in the UI.

### LINKED ISSUE #11146: Collection Edit: Support Banner Update
Add functionality to the collection edit modal allowing users to update the collection banner. Ensure that the updated banner is reflected on-chain and displayed correctly in the UI.

### LINKED ISSUE #11147: Collection Edit: Support Supply Update
Add functionality to the collection edit modal allowing users to update the collection supply. Ensure that any changes to the supply are correctly enforced and reflected on-chain.

### PR DESCRIPTION
**Pull request by hassnian**
## PR Type

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring

## Context

 not closing some issues since some UI/UX changes may be needed

- [x] fix: use banner from metadata in collection card
- [x] Closes #11099
- [x] Closes #5286
- [x] Closes #11146
- [x] Closes #11145
- [x] Closes #11147
- [x] Ref #11142
- [x] Ref #11153
- [x] Ref #11156 
- [x] Ref #11154
- [x]  Ref #11155

## Screenshot 📸

- [x] My fix has changed **something** on UI; 

### Edit NFT
CleanShot 2024-11-05 at 11 45 44@2x

### Edit Collection 
CleanShot 2024-10-30 at 18 42 03@2x

### Refresh Collection Nfts Metadata
CleanShot 2024-10-30 at 18 42 27@2x

### CODE REVIEW

File: composables/transaction/transactionUpdateCollection.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  isLoading.value = true
+  status.value = 'loader.ipfs'
+
+  const metadata = await constructMeta(item)
+
+  const collectionId = item.collectionId.toString()
+
+  const args = [
+    api.tx.nfts.setCollectionMetadata(collectionId, metadata),
+    api.tx.nfts.setCollectionMaxSupply(collectionId, item.collection.max),
```
Comment (vikiival) [inline composables/transaction/transactionUpdateCollection.ts:69]: if supply does not change, I would skip the call

**Comment by vikiival** [comment]
Amazing ❤️❤️
I am just thinking how to add 
- #11143
- #11144

also I would make the modal wider WDYT @exezbcz

File: composables/transaction/transactionUpdateCollection.ts (hunk 1 above)
Comment (hassnian) [reply composables/transaction/transactionUpdateCollection.ts]: done

**Comment by exezbcz** [comment]
>also I would make the modal wider WDYT @exezbcz

I would keep the width the same for the consistency

After adding more features, if we conclude it needs to be wider, we can change that.

Anyway, good job @hassnian !

**Comment by exezbcz** [comment]
will editing an nft create on chain event that will be in the activity tab?

File: components/collection/CollectionCard.vue [hunk 2]
```diff
@@ -55,17 +55,21 @@ const props = defineProps<{
...
 
 const { urlPrefix } = usePrefix()
 
 const getImageFromMetadata = async (collectionMetadata: string) => {
   isLoadingMeta.value = true
 
-  const metadata = (await processSingleMetadata(
-    collectionMetadata,
-  )) as TokenMetadata
+  const metadata = await processSingleMetadata<TokenMetadata>(collectionMetadata)
```
Comment (preschian) [inline components/collection/CollectionCard.vue:65]: update types? or using another types

**Review by preschian — APPROVED** [review]
great pr 🚀

File: components/collection/CollectionCard.vue (hunk 2 above)
Comment (hassnian) [reply components/collection/CollectionCard.vue:65]: the type did not change just using generic types

Comment (preschian) [reply components/collection/CollectionCard.vue:65]: `metadata.banner` does not exists in this type, no?

**Comment by hassnian** [comment]
> will editing an nft create on chain event that will be in the activity tab?


atm no, the indexer does not have an UPDATE_METADATA  event.

File: components/collection/CollectionCard.vue (hunk 2 above)
Comment (hassnian) [reply components/collection/CollectionCard.vue:65]: oh, yes this type doesn't have banner, should we update the `hyperdata`  type or create a local one

File: types/index.ts [hunk 3]
```diff
@@ -14,7 +14,7 @@ export interface Metadata {
   thumbnailUri?: string
   mediaUri?: string
   chain?: Prefix
-  meta?: TokenMetadata
+  meta?: T
 }
 
 export interface NFTMetadata extends Metadata {
```
Comment (vikiival) [inline types/index.ts:20]: Is this used somewhere?

File: composables/transaction/transactionUpdateCollection.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    name,
+    description,
+    image,
+    undefined,
+    attributes,
+    undefined,
+    type,
+  )
+
+  // TODO: add banner field to minimark https://github.com/kodadot/packages/blob/main/minimark/src/utils/metadata.ts#L4
```
Comment (vikiival) [inline composables/transaction/transactionUpdateCollection.ts:54]: I would rather do this

- https://github.com/kodadot/packages/issues/174

File: composables/transaction/types.ts [hunk 5]
```diff
@@ -241,7 +241,7 @@ export interface ActionMintCollection {
 
 export enum Collections {
```
Comment (vikiival) [inline composables/transaction/types.ts:242]: Why did you remove the set_max?

File: locales/en.json [hunk 6]
```diff
@@ -377,6 +377,31 @@
...
     "yourWalletNeeds": "Your wallet needs a minimum of {0} to claim this drop."
   },
+  "edit": {
+    "collection": {
+      "banner": {
+        "hint": "1440 x 560 size recommended",
+        "label": "Banner",
+        "message": "Represents your collection on its page and on collection card"
+      },
+      "drop": "Don't click, just drop!",
```
Comment (vikiival) [inline locales/en.json:387]: any reason for this one?

**Review by vikiival — COMMENTED** [review]
Some small things for clarification

File: locales/en.json (hunk 6 above)
Comment (hassnian) [reply locales/en.json:387]: from design 

CleanShot 2024-11-20 at 14 29 57@2x

File: composables/transaction/types.ts (hunk 5 above)
Comment (hassnian) [reply composables/transaction/types.ts:242]: `UPDATE_COLLECTION` is supposed to be the replacement

Comment (hassnian) [reply composables/transaction/types.ts:242]: `UPDATE_COLLECTION` is the replacement for that action

Comment (vikiival) [reply composables/transaction/types.ts:242]: yup I would keep it

**Comment by vikiival** [comment]
pls resolve conflicts

**Review by Jarsen136 — APPROVED** [review]

File: composables/transaction/types.ts (hunk 5 above)
Comment (hassnian) [reply composables/transaction/types.ts:242]: what about something like this [492f13b](https://github.com/kodadot/nft-gallery/pull/11171/commits/492f13b5557ea4b1f1cf55169ab24e258aa76d63), alternatively I can just a the old action `SET_MAX_SUPPLY` and just call that

**Comment by hassnian** [comment]
blocker #11187 

CleanShot 2024-11-26 at 11 07 49@2x
